packages/babel-plugin-transform-class-properties/src/index.js

```diff
@@ -32,106 +223,209 @@ export default function({ types: t }) {
+        node.value || classBody.scope.buildUndefinedNode(),
+      ]),
+    );
+  }
+
+  function buildPrivateClassPropertyLoose(ref, prop, classBody, nodes) {
+    const { key, value } = prop.node;
+    const { name } = key.id;
+    const { file } = classBody.hub;
+    const privateKey = classBody.scope.generateDeclaredUidIdentifier(name);
+
+    classBody.traverse(privateNameRemapperLoose, {
+      name,
+      privateKey,
+      base: file.addHelper("classPrivateFieldBase"),
+    });
+
+    nodes.push(
+      t.expressionStatement(
+        t.assignmentExpression(
+          "=",
+          privateKey,
+          t.callExpression(file.addHelper("classPrivateFieldKey"), [
+            t.stringLiteral(name),
+          ]),
+        ),
+      ),
+    );
+
+    return buildPrivateProperty({
+      REF: ref,
+      KEY: privateKey,
+      VALUE: value || prop.scope.buildUndefinedNode(),
+    });
+  }
 
   return {
     inherits: syntaxClassProperties,
 
     visitor: {
       Class(path, state) {
-        const buildClassProperty = state.opts.loose
-          ? buildClassPropertyLoose
-          : buildClassPropertySpec;
+        const buildPublicClassProperty = state.opts.loose
+          ? buildPublicClassPropertyLoose
+          : buildPublicClassPropertySpec;
+        const buildPrivateClassProperty = state.opts.loose
+          ? buildPrivateClassPropertyLoose
+          : buildPrivateClassPropertySpec;
         const isDerived = !!path.node.superClass;
-        let constructor;
-        const props = [];
+
+        const instanceProps = [];
+        const staticProps = [];
         const body = path.get("body");
+        const { scope } = path;
 
-        for (const path of body.get("body")) {
-          if (path.isClassProperty()) {
-            props.push(path);
-          } else if (path.isClassMethod({ kind: "constructor" })) {
-            constructor = path;
-          }
-        }
+        const propNames = {
+          publicProps: Object.create(null),
+          publicStaticProps: Object.create(null),
+          privateProps: Object.create(null),
+          instanceProps,
+          staticProps,
+          constructor: null,
+        };
+        collectPropertiesVisitor(body, propNames);
+        body.traverse(staticErrorVisitor, propNames);
 
-        if (!props.length) return;
+        if (!instanceProps.length && !staticProps.length) return;
 
-        const nodes = [];
         let ref;
-
         if (path.isClassExpression() || !path.node.id) {
           nameFunction(path);
-          ref = path.scope.generateUidIdentifier("class");
+          ref = scope.generateUidIdentifier("class");
         } else {
           // path.isClassDeclaration() && path.node.id
           ref = path.node.id;
         }
 
+        const nodes = [];
         let instanceBody = [];
 
-        for (const prop of props) {
-          const propNode = prop.node;
-          if (propNode.decorators && propNode.decorators.length > 0) continue;
```

Comment:
Seems like this line is there to allow the decorators transform to have its own implementation of the transform for fields. I'm not sure if this factoring is optimal forever, as decorators move towards the specification (cc @loganfsmyth ), but maybe the logic should be preserved for now to avoid breaking existing things. While you're at it, it'd be good to ensure that decorated private field declarations lead to a clean SyntaxError (ideally implemented as a check in the decorator transform, which would later be updated to support this case).